server/fleet/apple_mdm.go

```diff
@@ -322,6 +322,18 @@ func (cp MDMAppleConfigProfile) ValidateUserProvided() error {
 	return cp.Mobileconfig.ScreenPayloads()
 }
 
+// IsWithinGracePeriod returns true if the host is within the grace period for the profile.
+//
+// The grace period is defined as 1 hour after the profile was updated. It is checked against the
+// host's detail_updated_at timestamp to allow for the host to check in at least once before the
+// profile is considered failed. If the host is online, it should report detail queries hourly by
+// default. If the host is offline, it should report detail queries shortly after it comes back
+// online.
+func (cp MDMAppleConfigProfile) IsWithinGracePeriod(hostDetailUpdatedAt time.Time) bool {
```

Comment:
> I think there's a missing piece here, my understanding is that `cp.UpdatedAt` refers to when the configuration profile was last modified by the IT admin (as this function is called for each item of `expectedProfs` above) , but that could be a date way in the past making this always return `false`, right?

You're right. This implementation of the grace period only mitigates premature failures where a profile is added or updated after the host has been enrolled in MDM (team/no team). To also cover events where a host enrolls to MDM or is assigned to a different MDM team, we'd need to look at some additional data point that can tell us when that event occurred. As an aside, @mna brought up the event-sourcing pattern previously, which seems like it could potentially be applicable here too.


> Additionally, I'm worried about the following race condition, could you sanity check if I'm missing something?
> 
> 1. Host is offline
> 2. Host is switched to a different team, with different profiles (I'm using a team switch because I think makes the case more clear, but this can happen regardless)
> 3. Host checks in before MDM commands to install the new profiles / remove the old profiles have been delivered / run

This is a good illustration of the problem above. We need some place we can track host events. In this case, I think. we'd at a minimum want some timestamp(s) for the host's MDM enrollment event and host team assignment events. Maybe we could use the nano_enrollments table for this. It also includes a last_seen_at timestamp that is probably a better choice than the host.detail_updated_at. I don't know that we currently store a timestamp for team assignments apart from what might be in activities table (which would need to be extracted from the json detail).

Reply:
Some other thoughts:

There is a host activity feed on the roadmap, which will need to address the bigger picture issue of how we are storing host events.

There's a `host_updates` table that currently just records `software_updated_at`.

Reply:
@roperzh @gillespi314 

> 2. Host is switched to a different team, with different profiles (I'm using a team switch because I think makes the case more clear, but this can happen regardless)
> 3. Host checks in before MDM commands to install the new profiles / remove the old profiles have been delivered / run

Correct me if I'm wrong or misunderstood your point, but moving a host to a different team would set the status to `NULL` (pending), so the osquery checks would not apply (not verifying/verified yet)?

Reply:
> Correct me if I'm wrong or misunderstood your point, but moving a host to a different team would set the status to NULL (pending), so the osquery checks would not apply (not verifying/verified yet)?

@mna I think that might be the missing piece but I'm not sure, my understanding is that since right now we're getting the expected list of profiles from `mdm_apple_configuration_profiles`, and comparing that to whatever we've got from osquery, the checks will still apply.

I didn't read @gillespi314 answer in detail yet!

Reply:
@roperzh If you look at `setMDMProfilesFailedDB` (or `setMDMProfilesVerifiedDB`), it will only update host profiles entries that were in verified or verifying, ignoring the other ones (even if they were identified as missing/confirmed via the results from osquery).